// chainlink/dist/src/generated/CoordinatorInterfaceFactory.js
"use strict";
/* Generated by ts-generator ver. 0.0.8 */
/* tslint:disable */
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
class CoordinatorInterfaceFactory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.CoordinatorInterfaceFactory = CoordinatorInterfaceFactory;
const _abi = [
    {
        constant: false,
        inputs: [
            {
                components: [
                    {
                        name: "payment",
                        type: "uint256"
                    },
                    {
                        name: "expiration",
                        type: "uint256"
                    },
                    {
                        name: "endAt",
                        type: "uint256"
                    },
                    {
                        name: "oracles",
                        type: "address[]"
                    },
                    {
                        name: "requestDigest",
                        type: "bytes32"
                    },
                    {
                        name: "aggregator",
                        type: "address"
                    },
                    {
                        name: "aggInitiateJobSelector",
                        type: "bytes4"
                    },
                    {
                        name: "aggFulfillSelector",
                        type: "bytes4"
                    }
                ],
                name: "_agreement",
                type: "tuple"
            },
            {
                components: [
                    {
                        name: "vs",
                        type: "uint8[]"
                    },
                    {
                        name: "rs",
                        type: "bytes32[]"
                    },
                    {
                        name: "ss",
                        type: "bytes32[]"
                    }
                ],
                name: "_signatures",
                type: "tuple"
            }
        ],
        name: "initiateServiceAgreement",
        outputs: [
            {
                name: "",
                type: "bytes32"
            }
        ],
        payable: false,
        stateMutability: "nonpayable",
        type: "function"
    },
    {
        constant: false,
        inputs: [
            {
                name: "_requestId",
                type: "bytes32"
            },
            {
                name: "_aggregatorArgs",
                type: "bytes32"
            }
        ],
        name: "fulfillOracleRequest",
        outputs: [
            {
                name: "",
                type: "bool"
            }
        ],
        payable: false,
        stateMutability: "nonpayable",
        type: "function"
    }
];
