// set-protocol-contracts/dist/utils/contract_logs/exchangeIssuanceModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function LogExchangeIssue(setAddress, callerAddress, quantity, sendTokens, sendTokenAmounts, contractAddress) {
    return [{
            event: 'LogExchangeIssue',
            address: contractAddress,
            args: {
                setAddress: setAddress,
                callerAddress: callerAddress,
                quantity: quantity,
                sendTokens: sendTokens,
                sendTokenAmounts: sendTokenAmounts,
            },
        }];
}
exports.LogExchangeIssue = LogExchangeIssue;
function LogExchangeRedeem(setAddress, callerAddress, quantity, receiveTokens, receiveTokenAmounts, contractAddress) {
    return [{
            event: 'LogExchangeRedeem',
            address: contractAddress,
            args: {
                setAddress: setAddress,
                callerAddress: callerAddress,
                quantity: quantity,
                receiveTokens: receiveTokens,
                receiveTokenAmounts: receiveTokenAmounts,
            },
        }];
}
exports.LogExchangeRedeem = LogExchangeRedeem;
