// set-protocol-contracts/dist/utils/contract_logs/rebalancingSetIssuanceModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function LogRebalancingSetIssue(rebalancingSetAddress, callerAddress, rebalancingSetQuantity, contractAddress) {
    return [{
            event: 'LogRebalancingSetIssue',
            address: contractAddress,
            args: {
                rebalancingSetAddress: rebalancingSetAddress,
                callerAddress: callerAddress,
                rebalancingSetQuantity: rebalancingSetQuantity,
            },
        }];
}
exports.LogRebalancingSetIssue = LogRebalancingSetIssue;
function LogRebalancingSetRedeem(rebalancingSetAddress, callerAddress, rebalancingSetQuantity, contractAddress) {
    return [{
            event: 'LogRebalancingSetRedeem',
            address: contractAddress,
            args: {
                rebalancingSetAddress: rebalancingSetAddress,
                callerAddress: callerAddress,
                rebalancingSetQuantity: rebalancingSetQuantity,
            },
        }];
}
exports.LogRebalancingSetRedeem = LogRebalancingSetRedeem;
