// set-protocol-contracts/dist/utils/contract_logs/rebalancingSetExchangeIssuanceModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function LogPayableExchangeIssue(rebalancingSetAddress, callerAddress, paymentTokenAddress, rebalancingSetQuantity, paymentTokenReturned, contractAddress) {
    return [{
            event: 'LogPayableExchangeIssue',
            address: contractAddress,
            args: {
                rebalancingSetAddress: rebalancingSetAddress,
                callerAddress: callerAddress,
                paymentTokenAddress: paymentTokenAddress,
                rebalancingSetQuantity: rebalancingSetQuantity,
                paymentTokenReturned: paymentTokenReturned,
            },
        }];
}
exports.LogPayableExchangeIssue = LogPayableExchangeIssue;
function LogPayableExchangeRedeem(rebalancingSetAddress, callerAddress, outputTokenAddress, rebalancingSetQuantity, outputTokenQuantity, contractAddress) {
    return [{
            event: 'LogPayableExchangeRedeem',
            address: contractAddress,
            args: {
                rebalancingSetAddress: rebalancingSetAddress,
                callerAddress: callerAddress,
                outputTokenAddress: outputTokenAddress,
                rebalancingSetQuantity: rebalancingSetQuantity,
                outputTokenQuantity: outputTokenQuantity,
            },
        }];
}
exports.LogPayableExchangeRedeem = LogPayableExchangeRedeem;
