// set-protocol-contracts/dist/utils/contract_logs/rebalancingSetToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getExpectedTransferLog(from, to, value, contractAddress) {
    return [{
            event: 'Transfer',
            address: contractAddress,
            args: {
                from: from,
                to: to,
                value: value,
            },
        }];
}
exports.getExpectedTransferLog = getExpectedTransferLog;
function getExpectedNewManagerAddedLog(newManager, oldManager, contractAddress) {
    return [{
            event: 'NewManagerAdded',
            address: contractAddress,
            args: {
                newManager: newManager,
                oldManager: oldManager,
            },
        }];
}
exports.getExpectedNewManagerAddedLog = getExpectedNewManagerAddedLog;
function getExpectedRebalanceProposedLog(nextSet, auctionLibrary, proposalPeriodEndTime, contractAddress) {
    return [{
            event: 'RebalanceProposed',
            address: contractAddress,
            args: {
                nextSet: nextSet,
                auctionLibrary: auctionLibrary,
                proposalPeriodEndTime: proposalPeriodEndTime,
            },
        }];
}
exports.getExpectedRebalanceProposedLog = getExpectedRebalanceProposedLog;
function getExpectedRebalanceStartedLog(oldSet, newSet, contractAddress) {
    return [{
            event: 'RebalanceStarted',
            address: contractAddress,
            args: {
                oldSet: oldSet,
                newSet: newSet,
            },
        }];
}
exports.getExpectedRebalanceStartedLog = getExpectedRebalanceStartedLog;
