// chainlink/v0.5/dist/src/generated/LinkTokenReceiverFactory.js
"use strict";
/* Generated by ts-generator ver. 0.0.8 */
/* tslint:disable */
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
class LinkTokenReceiverFactory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.LinkTokenReceiverFactory = LinkTokenReceiverFactory;
const _abi = [
    {
        constant: true,
        inputs: [],
        name: "getChainlinkToken",
        outputs: [
            {
                name: "",
                type: "address"
            }
        ],
        payable: false,
        stateMutability: "view",
        type: "function"
    },
    {
        constant: false,
        inputs: [
            {
                name: "_sender",
                type: "address"
            },
            {
                name: "_amount",
                type: "uint256"
            },
            {
                name: "_data",
                type: "bytes"
            }
        ],
        name: "onTokenTransfer",
        outputs: [],
        payable: false,
        stateMutability: "nonpayable",
        type: "function"
    }
];
