// @0xproject/json-schemas/lib/schemas/relayer_api_orders_channel_subscribe_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.relayerApiOrdersChannelSubscribeSchema = {
    id: '/relayerApiOrdersChannelSubscribeSchema',
    type: 'object',
    properties: {
        type: { enum: ['subscribe'] },
        channel: { enum: ['orders'] },
        requestId: { type: 'string' },
        payload: { $ref: '/relayerApiOrdersChannelSubscribePayload' },
    },
    required: ['type', 'channel', 'requestId'],
};
exports.relayerApiOrdersChannelSubscribePayload = {
    id: '/relayerApiOrdersChannelSubscribePayload',
    type: 'object',
    properties: {
        makerAssetProxyId: { $ref: '/hexSchema' },
        takerAssetProxyId: { $ref: '/hexSchema' },
        networkId: { type: 'number' },
        makerAssetAddress: { $ref: '/addressSchema' },
        takerAssetAddress: { $ref: '/addressSchema' },
        makerAssetData: { $ref: '/hexSchema' },
        takerAssetData: { $ref: '/hexSchema' },
        traderAssetData: { $ref: '/hexSchema' },
    },
};
