// @0xproject/json-schemas/lib/schemas/relayer_api_asset_pairs_response_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.relayerApiAssetDataPairsResponseSchema = {
    id: '/relayerApiAssetDataPairsResponseSchema',
    type: 'object',
    allOf: [
        { $ref: '/paginatedCollectionSchema' },
        {
            properties: {
                records: { $ref: '/relayerApiAssetDataPairsSchema' },
            },
            required: ['records'],
        },
    ],
};
exports.relayerApiAssetDataPairsSchema = {
    id: '/relayerApiAssetDataPairsSchema',
    type: 'array',
    items: {
        properties: {
            assetDataA: { $ref: '/relayerApiAssetDataTradeInfoSchema' },
            assetDataB: { $ref: '/relayerApiAssetDataTradeInfoSchema' },
        },
        required: ['assetDataA', 'assetDataB'],
        type: 'object',
    },
};
exports.relayerApiAssetDataTradeInfoSchema = {
    id: '/relayerApiAssetDataTradeInfoSchema',
    type: 'object',
    properties: {
        assetData: { $ref: '/hexSchema' },
        minAmount: { $ref: '/numberSchema' },
        maxAmount: { $ref: '/numberSchema' },
        precision: { type: 'number' },
    },
    required: ['assetData'],
};
