// set-protocol-contracts/dist/utils/helpers/libraryMockHelper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var contracts_1 = require("../contracts");
var bignumber_js_1 = require("bignumber.js");
var web3Helper_1 = require("../web3Helper");
var constants_1 = require("../constants");
var CoreIssuanceLibrary = web3Helper_1.importArtifactsFromSource('CoreIssuanceLibrary');
var CoreIssuanceLibraryMock = web3Helper_1.importArtifactsFromSource('CoreIssuanceLibraryMock');
var ERC20Wrapper = web3Helper_1.importArtifactsFromSource('ERC20Wrapper');
var ERC20WrapperMock = web3Helper_1.importArtifactsFromSource('ERC20WrapperMock');
var ExchangeIssuanceLibraryMock = web3Helper_1.importArtifactsFromSource('ExchangeIssuanceLibraryMock');
var PlaceBidMock = web3Helper_1.importArtifactsFromSource('PlaceBidMock');
var RebalanceMock = web3Helper_1.importArtifactsFromSource('RebalanceMock');
var RebalancingSetIssuanceMock = web3Helper_1.importArtifactsFromSource('RebalancingSetIssuanceMock');
var SetMathMock = web3Helper_1.importArtifactsFromSource('SetMathMock');
var SetTokenLibrary = web3Helper_1.importArtifactsFromSource('SetTokenLibrary');
var SetTokenLibraryMock = web3Helper_1.importArtifactsFromSource('SetTokenLibraryMock');
var RebalanceStateSetTokenMock = web3Helper_1.importArtifactsFromSource('RebalanceStateSetTokenMock');
var SetUSDValuationMock = web3Helper_1.importArtifactsFromSource('SetUSDValuationMock');
var TokenFlushMock = web3Helper_1.importArtifactsFromSource('TokenFlushMock');
var ZeroExOrderLibraryMock = web3Helper_1.importArtifactsFromSource('ZeroExOrderLibraryMock');
var LibraryMockHelper = /** @class */ (function () {
    function LibraryMockHelper(contractOwnerAddress) {
        this._contractOwnerAddress = contractOwnerAddress;
    }
    /* ============ Deployment ============ */
    LibraryMockHelper.prototype.deployCoreIssuanceLibraryAsync = function (from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var truffleCoreIssuanceLibraryMock;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, web3Helper_1.linkLibrariesToDeploy(CoreIssuanceLibraryMock, [CoreIssuanceLibrary], this._contractOwnerAddress)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, CoreIssuanceLibraryMock.new(web3Helper_1.txnFrom(from))];
                    case 2:
                        truffleCoreIssuanceLibraryMock = _a.sent();
                        return [2 /*return*/, new contracts_1.CoreIssuanceLibraryMockContract(web3Helper_1.getContractInstance(truffleCoreIssuanceLibraryMock), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deployExchangeIssuanceLibraryAsync = function (from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var exchangeIssuanceMockContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, ExchangeIssuanceLibraryMock.new(web3Helper_1.txnFrom(from))];
                    case 1:
                        exchangeIssuanceMockContract = _a.sent();
                        return [2 /*return*/, new contracts_1.ExchangeIssuanceLibraryMockContract(web3Helper_1.getContractInstance(exchangeIssuanceMockContract), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deployTokenFlushMockAsync = function (core, vault, transferProxy, from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var tokenFlushMockContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, web3Helper_1.linkLibrariesToDeploy(TokenFlushMock, [ERC20Wrapper], this._contractOwnerAddress)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, TokenFlushMock.new(core.address, vault.address, transferProxy.address, web3Helper_1.txnFrom(from))];
                    case 2:
                        tokenFlushMockContract = _a.sent();
                        return [2 /*return*/, new contracts_1.TokenFlushMockContract(web3Helper_1.getContractInstance(tokenFlushMockContract), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deployRebalanceMockAsync = function (from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var rebalanceMockContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, RebalanceMock.new(web3Helper_1.txnFrom(from))];
                    case 1:
                        rebalanceMockContract = _a.sent();
                        return [2 /*return*/, new contracts_1.RebalanceMockContract(web3Helper_1.getContractInstance(rebalanceMockContract), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deployRebalanceStateSetTokenMockAsync = function (combinedTokenArray, inflowArray, outflowArray, name, symbol, decimals, from) {
        if (name === void 0) { name = 'Rebalancing Set Token'; }
        if (symbol === void 0) { symbol = 'RBSET'; }
        if (decimals === void 0) { decimals = new bignumber_js_1.BigNumber(18); }
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var rebalanceStateSetTokenMock;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, RebalanceStateSetTokenMock.new(name, symbol, decimals, combinedTokenArray, inflowArray, outflowArray, web3Helper_1.txnFrom(from))];
                    case 1:
                        rebalanceStateSetTokenMock = _a.sent();
                        return [2 /*return*/, new contracts_1.RebalanceStateSetTokenMockContract(web3Helper_1.getContractInstance(rebalanceStateSetTokenMock), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deployRebalancingSetIssuanceMockAsync = function (core, vault, from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var rebalancingSetIssuanceMockContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, web3Helper_1.linkLibrariesToDeploy(RebalancingSetIssuanceMock, [ERC20Wrapper], this._contractOwnerAddress)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, RebalancingSetIssuanceMock.new(core.address, vault.address, web3Helper_1.txnFrom(from))];
                    case 2:
                        rebalancingSetIssuanceMockContract = _a.sent();
                        return [2 /*return*/, new contracts_1.RebalancingSetIssuanceMockContract(web3Helper_1.getContractInstance(rebalancingSetIssuanceMockContract), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deployERC20WrapperLibraryAsync = function (from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var erc20WrapperMockContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, ERC20WrapperMock.new(web3Helper_1.txnFrom(from))];
                    case 1:
                        erc20WrapperMockContract = _a.sent();
                        return [2 /*return*/, new contracts_1.ERC20WrapperMockContract(web3Helper_1.getContractInstance(erc20WrapperMockContract), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deploySetMathAsync = function (from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var setMathMockContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, SetMathMock.new(web3Helper_1.txnFrom(from))];
                    case 1:
                        setMathMockContract = _a.sent();
                        return [2 /*return*/, new contracts_1.SetMathMockContract(web3Helper_1.getContractInstance(setMathMockContract), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deploySetTokenLibraryAsync = function (from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var setTokenLibraryMockContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, web3Helper_1.linkLibrariesToDeploy(SetTokenLibraryMock, [SetTokenLibrary], this._contractOwnerAddress)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, SetTokenLibraryMock.new(web3Helper_1.txnFrom(from))];
                    case 2:
                        setTokenLibraryMockContract = _a.sent();
                        return [2 /*return*/, new contracts_1.SetTokenLibraryMockContract(web3Helper_1.getContractInstance(setTokenLibraryMockContract), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deploySetUSDValuationMockAsync = function (from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var setValuationMockContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, SetUSDValuationMock.new(web3Helper_1.txnFrom(from))];
                    case 1:
                        setValuationMockContract = _a.sent();
                        return [2 /*return*/, new contracts_1.SetUSDValuationMockContract(web3Helper_1.getContractInstance(setValuationMockContract), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deployZeroExOrderLibraryAsync = function (from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var zeroExExchangeWrapperInstance;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, ZeroExOrderLibraryMock.new(web3Helper_1.txnFrom(from))];
                    case 1:
                        zeroExExchangeWrapperInstance = _a.sent();
                        return [2 /*return*/, new contracts_1.ZeroExOrderLibraryMockContract(web3Helper_1.getContractInstance(zeroExExchangeWrapperInstance), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.deployPlaceBidMockAsync = function (from) {
        if (from === void 0) { from = this._contractOwnerAddress; }
        return __awaiter(this, void 0, void 0, function () {
            var placeBidMockContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, PlaceBidMock.new(web3Helper_1.txnFrom(from))];
                    case 1:
                        placeBidMockContract = _a.sent();
                        return [2 /*return*/, new contracts_1.PlaceBidMockContract(web3Helper_1.getContractInstance(placeBidMockContract), web3Helper_1.txnFrom(from))];
                }
            });
        });
    };
    LibraryMockHelper.prototype.ceilLog10 = function (value) {
        var valueNum = value.toNumber();
        if (valueNum == 1)
            return constants_1.ZERO;
        var x = valueNum - 1;
        var result = 0;
        if (x >= 10000000000000000000000000000000000000000000000000000000000000000) {
            x /= 10000000000000000000000000000000000000000000000000000000000000000;
            result += 64;
        }
        if (x >= 100000000000000000000000000000000) {
            x /= 100000000000000000000000000000000;
            result += 32;
        }
        if (x >= 10000000000000000) {
            x /= 10000000000000000;
            result += 16;
        }
        if (x >= 100000000) {
            x /= 100000000;
            result += 8;
        }
        if (x >= 10000) {
            x /= 10000;
            result += 4;
        }
        if (x >= 100) {
            x /= 100;
            result += 2;
        }
        if (x >= 10) {
            x /= 10;
            result += 1;
        }
        return new bignumber_js_1.BigNumber(result + 1);
    };
    return LibraryMockHelper;
}());
exports.LibraryMockHelper = LibraryMockHelper;
