// set-protocol-contracts/dist/utils/auction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var bignumber_js_1 = require("bignumber.js");
function getLinearAuction(input) {
    var _a = input.auction, minimumBid = _a.minimumBid, startTime = _a.startTime, startingCurrentSets = _a.startingCurrentSets, remainingCurrentSets = _a.remainingCurrentSets, combinedCurrentSetUnits = _a.combinedCurrentSetUnits, combinedNextSetUnits = _a.combinedNextSetUnits, maxNaturalUnit = _a.maxNaturalUnit;
    return {
        auction: {
            maxNaturalUnit: new bignumber_js_1.BigNumber(maxNaturalUnit),
            minimumBid: new bignumber_js_1.BigNumber(minimumBid),
            startTime: new bignumber_js_1.BigNumber(startTime),
            startingCurrentSets: new bignumber_js_1.BigNumber(startingCurrentSets),
            remainingCurrentSets: new bignumber_js_1.BigNumber(remainingCurrentSets),
            combinedTokenArray: input.auction.combinedTokenArray,
            combinedCurrentSetUnits: combinedCurrentSetUnits.map(function (v) { return new bignumber_js_1.BigNumber(v); }),
            combinedNextSetUnits: combinedNextSetUnits.map(function (v) { return new bignumber_js_1.BigNumber(v); }),
        },
        endTime: new bignumber_js_1.BigNumber(input.endTime),
        startPrice: new bignumber_js_1.BigNumber(input.startPrice),
        endPrice: new bignumber_js_1.BigNumber(input.endPrice),
    };
}
exports.getLinearAuction = getLinearAuction;
