// set-protocol-contracts/dist/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var bignumber_js_1 = require("bignumber.js");
var units_1 = require("../utils/units");
exports.AUCTION_TIME_INCREMENT = new bignumber_js_1.BigNumber(30); // Unix seconds
exports.AUCTION_CURVE_DENOMINATOR = units_1.ether(1);
exports.DEFAULT_AUCTION_PRICE_NUMERATOR = new bignumber_js_1.BigNumber(1374);
exports.DEFAULT_AUCTION_PRICE_DIVISOR = new bignumber_js_1.BigNumber(1000);
exports.DEFAULT_GAS = 19000000;
exports.DEFAULT_MOCK_TOKEN_DECIMALS = 18;
exports.DEFAULT_REBALANCE_START_PRICE = new bignumber_js_1.BigNumber(500);
exports.DEFAULT_REBALANCE_TIME_TO_PIVOT = new bignumber_js_1.BigNumber(100000);
exports.DEFAULT_REBALANCING_NATURAL_UNIT = new bignumber_js_1.BigNumber(Math.pow(10, 6));
exports.DEFAULT_REBALANCING_MINIMUM_NATURAL_UNIT = new bignumber_js_1.BigNumber(Math.pow(10, 4));
exports.DEFAULT_REBALANCING_MAXIMUM_NATURAL_UNIT = new bignumber_js_1.BigNumber(Math.pow(10, 12));
exports.DEFAULT_UNIT_SHARES = new bignumber_js_1.BigNumber(Math.pow(10, 6));
exports.DEPLOYED_TOKEN_QUANTITY = units_1.ether(100000000000);
exports.EMPTY_BYTESTRING = '0x00';
exports.KYBER_RESERVE_CONFIGURED_RATE = new bignumber_js_1.BigNumber('321556325999999997');
exports.NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.ONE = new bignumber_js_1.BigNumber(1);
exports.ONE_HUNDRED = new bignumber_js_1.BigNumber(100);
exports.ONE_DAY_IN_SECONDS = new bignumber_js_1.BigNumber(86400);
exports.ONE_HOUR_IN_SECONDS = new bignumber_js_1.BigNumber(3600);
exports.ONE_YEAR_IN_SECONDS = exports.ONE_DAY_IN_SECONDS.mul(365.25);
exports.SCALE_FACTOR = units_1.ether(1);
exports.STANDARD_COMPONENT_UNIT = units_1.ether(1);
exports.STANDARD_NATURAL_UNIT = units_1.ether(1);
exports.STANDARD_QUANTITY_ISSUED = units_1.ether(10);
exports.UNLIMITED_ALLOWANCE_IN_BASE_UNITS = new bignumber_js_1.BigNumber(2).pow(256).minus(1);
exports.ZERO = new bignumber_js_1.BigNumber(0);
exports.ZERO_BYTES = '0x0000000000000000000000000000000000000000000000000000000000000000';
