// set-protocol-contracts/dist/utils/contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var address_to_address_white_list_1 = require("../types/generated/address_to_address_white_list");
exports.AddressToAddressWhiteListContract = address_to_address_white_list_1.AddressToAddressWhiteListContract;
var auction_getters_mock_1 = require("../types/generated/auction_getters_mock");
exports.AuctionGettersMockContract = auction_getters_mock_1.AuctionGettersMockContract;
var auction_mock_1 = require("../types/generated/auction_mock");
exports.AuctionMockContract = auction_mock_1.AuctionMockContract;
var authorizable_1 = require("../types/generated/authorizable");
exports.AuthorizableContract = authorizable_1.AuthorizableContract;
var bad_token_mock_1 = require("../types/generated/bad_token_mock");
exports.BadTokenMockContract = bad_token_mock_1.BadTokenMockContract;
var bad_c_token_mock_1 = require("../types/generated/bad_c_token_mock");
exports.BadCTokenMockContract = bad_c_token_mock_1.BadCTokenMockContract;
var base_contract_1 = require("../types/base_contract");
exports.BaseContract = base_contract_1.BaseContract;
var bytes32_library_1 = require("../types/generated/bytes32_library");
exports.Bytes32LibraryContract = bytes32_library_1.Bytes32LibraryContract;
var common_validations_library_1 = require("../types/generated/common_validations_library");
exports.CommonValidationsLibraryContract = common_validations_library_1.CommonValidationsLibraryContract;
var constant_auction_price_curve_1 = require("../types/generated/constant_auction_price_curve");
exports.ConstantAuctionPriceCurveContract = constant_auction_price_curve_1.ConstantAuctionPriceCurveContract;
var core_1 = require("../types/generated/core");
exports.CoreContract = core_1.CoreContract;
var core_issuance_library_1 = require("../types/generated/core_issuance_library");
exports.CoreIssuanceLibraryContract = core_issuance_library_1.CoreIssuanceLibraryContract;
var core_issuance_library_mock_1 = require("../types/generated/core_issuance_library_mock");
exports.CoreIssuanceLibraryMockContract = core_issuance_library_mock_1.CoreIssuanceLibraryMockContract;
var core_mock_1 = require("../types/generated/core_mock");
exports.CoreMockContract = core_mock_1.CoreMockContract;
var c_token_exchange_issuance_module_1 = require("../types/generated/c_token_exchange_issuance_module");
exports.CTokenExchangeIssuanceModuleContract = c_token_exchange_issuance_module_1.CTokenExchangeIssuanceModuleContract;
var erc20_detailed_1 = require("../types/generated/erc20_detailed");
exports.ERC20DetailedContract = erc20_detailed_1.ERC20DetailedContract;
var erc20_wrapper_1 = require("../types/generated/erc20_wrapper");
exports.ERC20WrapperContract = erc20_wrapper_1.ERC20WrapperContract;
var erc20_wrapper_mock_1 = require("../types/generated/erc20_wrapper_mock");
exports.ERC20WrapperMockContract = erc20_wrapper_mock_1.ERC20WrapperMockContract;
var exchange_issuance_library_1 = require("../types/generated/exchange_issuance_library");
exports.ExchangeIssuanceLibraryContract = exchange_issuance_library_1.ExchangeIssuanceLibraryContract;
var exchange_issuance_library_mock_1 = require("../types/generated/exchange_issuance_library_mock");
exports.ExchangeIssuanceLibraryMockContract = exchange_issuance_library_mock_1.ExchangeIssuanceLibraryMockContract;
var exchange_issuance_module_1 = require("../types/generated/exchange_issuance_module");
exports.ExchangeIssuanceModuleContract = exchange_issuance_module_1.ExchangeIssuanceModuleContract;
var fail_auction_library_1 = require("../types/generated/fail_auction_library");
exports.FailAuctionLibraryContract = fail_auction_library_1.FailAuctionLibraryContract;
var factory_utils_library_1 = require("../types/generated/factory_utils_library");
exports.FactoryUtilsLibraryContract = factory_utils_library_1.FactoryUtilsLibraryContract;
var fee_calculator_mock_1 = require("../types/generated/fee_calculator_mock");
exports.FeeCalculatorMockContract = fee_calculator_mock_1.FeeCalculatorMockContract;
var fixed_fee_calculator_mock_1 = require("../types/generated/fixed_fee_calculator_mock");
exports.FixedFeeCalculatorMockContract = fixed_fee_calculator_mock_1.FixedFeeCalculatorMockContract;
var fixed_fee_calculator_1 = require("../types/generated/fixed_fee_calculator");
exports.FixedFeeCalculatorContract = fixed_fee_calculator_1.FixedFeeCalculatorContract;
var i_auction_price_curve_1 = require("../types/generated/i_auction_price_curve");
exports.IAuctionPriceCurveContract = i_auction_price_curve_1.IAuctionPriceCurveContract;
var invalid_return_token_mock_1 = require("../types/generated/invalid_return_token_mock");
exports.InvalidReturnTokenMockContract = invalid_return_token_mock_1.InvalidReturnTokenMockContract;
var kyber_network_wrapper_1 = require("../types/generated/kyber_network_wrapper");
exports.KyberNetworkWrapperContract = kyber_network_wrapper_1.KyberNetworkWrapperContract;
var linear_auction_liquidator_1 = require("../types/generated/linear_auction_liquidator");
exports.LinearAuctionLiquidatorContract = linear_auction_liquidator_1.LinearAuctionLiquidatorContract;
var linear_auction_mock_1 = require("../types/generated/linear_auction_mock");
exports.LinearAuctionMockContract = linear_auction_mock_1.LinearAuctionMockContract;
var linear_auction_price_curve_1 = require("../types/generated/linear_auction_price_curve");
exports.LinearAuctionPriceCurveContract = linear_auction_price_curve_1.LinearAuctionPriceCurveContract;
var liquidator_mock_1 = require("../types/generated/liquidator_mock");
exports.LiquidatorMockContract = liquidator_mock_1.LiquidatorMockContract;
var liquidator_proxy_1 = require("../types/generated/liquidator_proxy");
exports.LiquidatorProxyContract = liquidator_proxy_1.LiquidatorProxyContract;
var liquidator_utils_mock_1 = require("../types/generated/liquidator_utils_mock");
exports.LiquidatorUtilsMockContract = liquidator_utils_mock_1.LiquidatorUtilsMockContract;
var no_decimal_token_mock_1 = require("../types/generated/no_decimal_token_mock");
exports.NoDecimalTokenMockContract = no_decimal_token_mock_1.NoDecimalTokenMockContract;
var no_xfer_return_token_mock_1 = require("../types/generated/no_xfer_return_token_mock");
exports.NoXferReturnTokenMockContract = no_xfer_return_token_mock_1.NoXferReturnTokenMockContract;
var oracle_white_list_1 = require("../types/generated/oracle_white_list");
exports.OracleWhiteListContract = oracle_white_list_1.OracleWhiteListContract;
var performance_fee_calculator_1 = require("../types/generated/performance_fee_calculator");
exports.PerformanceFeeCalculatorContract = performance_fee_calculator_1.PerformanceFeeCalculatorContract;
var place_bid_library_1 = require("../types/generated/place_bid_library");
exports.PlaceBidLibraryContract = place_bid_library_1.PlaceBidLibraryContract;
var place_bid_mock_1 = require("../types/generated/place_bid_mock");
exports.PlaceBidMockContract = place_bid_mock_1.PlaceBidMockContract;
var propose_library_1 = require("../types/generated/propose_library");
exports.ProposeLibraryContract = propose_library_1.ProposeLibraryContract;
var rebalance_auction_module_1 = require("../types/generated/rebalance_auction_module");
exports.RebalanceAuctionModuleContract = rebalance_auction_module_1.RebalanceAuctionModuleContract;
var rebalance_auction_module_mock_1 = require("../types/generated/rebalance_auction_module_mock");
exports.RebalanceAuctionModuleMockContract = rebalance_auction_module_mock_1.RebalanceAuctionModuleMockContract;
var rebalance_mock_1 = require("../types/generated/rebalance_mock");
exports.RebalanceMockContract = rebalance_mock_1.RebalanceMockContract;
var rebalancing_set_eth_bidder_1 = require("../types/generated/rebalancing_set_eth_bidder");
exports.RebalancingSetEthBidderContract = rebalancing_set_eth_bidder_1.RebalancingSetEthBidderContract;
var rebalancing_set_fee_mock_1 = require("../types/generated/rebalancing_set_fee_mock");
exports.RebalancingSetFeeMockContract = rebalancing_set_fee_mock_1.RebalancingSetFeeMockContract;
var rebalancing_set_c_token_bidder_1 = require("../types/generated/rebalancing_set_c_token_bidder");
exports.RebalancingSetCTokenBidderContract = rebalancing_set_c_token_bidder_1.RebalancingSetCTokenBidderContract;
var rebalancing_set_c_token_exchange_issuance_module_1 = require("../types/generated/rebalancing_set_c_token_exchange_issuance_module");
exports.RebalancingSetCTokenExchangeIssuanceModuleContract = rebalancing_set_c_token_exchange_issuance_module_1.RebalancingSetCTokenExchangeIssuanceModuleContract;
var rebalancing_set_c_token_issuance_module_1 = require("../types/generated/rebalancing_set_c_token_issuance_module");
exports.RebalancingSetCTokenIssuanceModuleContract = rebalancing_set_c_token_issuance_module_1.RebalancingSetCTokenIssuanceModuleContract;
var rebalancing_set_exchange_issuance_module_1 = require("../types/generated/rebalancing_set_exchange_issuance_module");
exports.RebalancingSetExchangeIssuanceModuleContract = rebalancing_set_exchange_issuance_module_1.RebalancingSetExchangeIssuanceModuleContract;
var rebalancing_set_issuance_module_1 = require("../types/generated/rebalancing_set_issuance_module");
exports.RebalancingSetIssuanceModuleContract = rebalancing_set_issuance_module_1.RebalancingSetIssuanceModuleContract;
var rebalancing_set_issuance_mock_1 = require("../types/generated/rebalancing_set_issuance_mock");
exports.RebalancingSetIssuanceMockContract = rebalancing_set_issuance_mock_1.RebalancingSetIssuanceMockContract;
var rebalancing_set_token_1 = require("../types/generated/rebalancing_set_token");
exports.RebalancingSetTokenContract = rebalancing_set_token_1.RebalancingSetTokenContract;
var rebalancing_set_token_v2_1 = require("../types/generated/rebalancing_set_token_v2");
exports.RebalancingSetTokenV2Contract = rebalancing_set_token_v2_1.RebalancingSetTokenV2Contract;
var rebalancing_set_token_v3_1 = require("../types/generated/rebalancing_set_token_v3");
exports.RebalancingSetTokenV3Contract = rebalancing_set_token_v3_1.RebalancingSetTokenV3Contract;
var rebalancing_set_token_factory_1 = require("../types/generated/rebalancing_set_token_factory");
exports.RebalancingSetTokenFactoryContract = rebalancing_set_token_factory_1.RebalancingSetTokenFactoryContract;
var rebalancing_set_token_v2_factory_1 = require("../types/generated/rebalancing_set_token_v2_factory");
exports.RebalancingSetTokenV2FactoryContract = rebalancing_set_token_v2_factory_1.RebalancingSetTokenV2FactoryContract;
var rebalancing_set_token_v3_factory_1 = require("../types/generated/rebalancing_set_token_v3_factory");
exports.RebalancingSetTokenV3FactoryContract = rebalancing_set_token_v3_factory_1.RebalancingSetTokenV3FactoryContract;
var settle_rebalance_library_1 = require("../types/generated/settle_rebalance_library");
exports.SettleRebalanceLibraryContract = settle_rebalance_library_1.SettleRebalanceLibraryContract;
var set_math_mock_1 = require("../types/generated/set_math_mock");
exports.SetMathMockContract = set_math_mock_1.SetMathMockContract;
var set_token_1 = require("../types/generated/set_token");
exports.SetTokenContract = set_token_1.SetTokenContract;
var set_token_factory_1 = require("../types/generated/set_token_factory");
exports.SetTokenFactoryContract = set_token_factory_1.SetTokenFactoryContract;
var set_token_library_1 = require("../types/generated/set_token_library");
exports.SetTokenLibraryContract = set_token_library_1.SetTokenLibraryContract;
var set_token_library_mock_1 = require("../types/generated/set_token_library_mock");
exports.SetTokenLibraryMockContract = set_token_library_mock_1.SetTokenLibraryMockContract;
var rebalance_state_set_token_mock_1 = require("../types/generated/rebalance_state_set_token_mock");
exports.RebalanceStateSetTokenMockContract = rebalance_state_set_token_mock_1.RebalanceStateSetTokenMockContract;
var set_u_s_d_valuation_mock_1 = require("../types/generated/set_u_s_d_valuation_mock");
exports.SetUSDValuationMockContract = set_u_s_d_valuation_mock_1.SetUSDValuationMockContract;
var social_trading_manager_mock_1 = require("../types/generated/social_trading_manager_mock");
exports.SocialTradingManagerMockContract = social_trading_manager_mock_1.SocialTradingManagerMockContract;
var standard_token_mock_1 = require("../types/generated/standard_token_mock");
exports.StandardTokenMockContract = standard_token_mock_1.StandardTokenMockContract;
var standard_token_with_fee_mock_1 = require("../types/generated/standard_token_with_fee_mock");
exports.StandardTokenWithFeeMockContract = standard_token_with_fee_mock_1.StandardTokenWithFeeMockContract;
var start_rebalance_library_1 = require("../types/generated/start_rebalance_library");
exports.StartRebalanceLibraryContract = start_rebalance_library_1.StartRebalanceLibraryContract;
var t_w_a_p_auction_mock_1 = require("../types/generated/t_w_a_p_auction_mock");
exports.TWAPAuctionMockContract = t_w_a_p_auction_mock_1.TWAPAuctionMockContract;
var t_w_a_p_auction_caller_1 = require("../types/generated/t_w_a_p_auction_caller");
exports.TWAPAuctionCallerContract = t_w_a_p_auction_caller_1.TWAPAuctionCallerContract;
var token_flush_mock_1 = require("../types/generated/token_flush_mock");
exports.TokenFlushMockContract = token_flush_mock_1.TokenFlushMockContract;
var time_lock_upgrade_1 = require("../types/generated/time_lock_upgrade");
exports.TimeLockUpgradeContract = time_lock_upgrade_1.TimeLockUpgradeContract;
var time_lock_upgrade_v2_1 = require("../types/generated/time_lock_upgrade_v2");
exports.TimeLockUpgradeV2Contract = time_lock_upgrade_v2_1.TimeLockUpgradeV2Contract;
var transfer_proxy_1 = require("../types/generated/transfer_proxy");
exports.TransferProxyContract = transfer_proxy_1.TransferProxyContract;
var t_w_a_p_auction_getters_mock_1 = require("../types/generated/t_w_a_p_auction_getters_mock");
exports.TWAPAuctionGettersMockContract = t_w_a_p_auction_getters_mock_1.TWAPAuctionGettersMockContract;
var t_w_a_p_liquidator_1 = require("../types/generated/t_w_a_p_liquidator");
exports.TWAPLiquidatorContract = t_w_a_p_liquidator_1.TWAPLiquidatorContract;
var two_asset_price_bounded_linear_auction_mock_1 = require("../types/generated/two_asset_price_bounded_linear_auction_mock");
exports.TwoAssetPriceBoundedLinearAuctionMockContract = two_asset_price_bounded_linear_auction_mock_1.TwoAssetPriceBoundedLinearAuctionMockContract;
var updatable_constant_auction_price_curve_1 = require("../types/generated/updatable_constant_auction_price_curve");
exports.UpdatableConstantAuctionPriceCurveContract = updatable_constant_auction_price_curve_1.UpdatableConstantAuctionPriceCurveContract;
var vault_1 = require("../types/generated/vault");
exports.VaultContract = vault_1.VaultContract;
var weth_mock_1 = require("../types/generated/weth_mock");
exports.WethMockContract = weth_mock_1.WethMockContract;
var white_list_1 = require("../types/generated/white_list");
exports.WhiteListContract = white_list_1.WhiteListContract;
var zero_ex_exchange_wrapper_1 = require("../types/generated/zero_ex_exchange_wrapper");
exports.ZeroExExchangeWrapperContract = zero_ex_exchange_wrapper_1.ZeroExExchangeWrapperContract;
var zero_ex_order_library_mock_1 = require("../types/generated/zero_ex_order_library_mock");
exports.ZeroExOrderLibraryMockContract = zero_ex_order_library_mock_1.ZeroExOrderLibraryMockContract;
