// set-protocol-oracles/dist/artifacts/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ChainlinkAggregatorMock_1 = require("./ts/ChainlinkAggregatorMock");
exports.ChainlinkAggregatorMock = ChainlinkAggregatorMock_1.ChainlinkAggregatorMock;
var ChainlinkOracleAdapter_1 = require("./ts/ChainlinkOracleAdapter");
exports.ChainlinkOracleAdapter = ChainlinkOracleAdapter_1.ChainlinkOracleAdapter;
var ConstantPriceOracle_1 = require("./ts/ConstantPriceOracle");
exports.ConstantPriceOracle = ConstantPriceOracle_1.ConstantPriceOracle;
var CTokenOracle_1 = require("./ts/CTokenOracle");
exports.CTokenOracle = CTokenOracle_1.CTokenOracle;
var DydxOracleAdapter_1 = require("./ts/DydxOracleAdapter");
exports.DydxOracleAdapter = DydxOracleAdapter_1.DydxOracleAdapter;
var DydxConstantPriceOracleMock_1 = require("./ts/DydxConstantPriceOracleMock");
exports.DydxConstantPriceOracleMock = DydxConstantPriceOracleMock_1.DydxConstantPriceOracleMock;
var EMAOracle_1 = require("./ts/EMAOracle");
exports.EMAOracle = EMAOracle_1.EMAOracle;
var FeedFactory_1 = require("./ts/FeedFactory");
exports.FeedFactory = FeedFactory_1.FeedFactory;
var HistoricalPriceFeed_1 = require("./ts/HistoricalPriceFeed");
exports.HistoricalPriceFeed = HistoricalPriceFeed_1.HistoricalPriceFeed;
var TwoAssetRatioOracle_1 = require("./ts/TwoAssetRatioOracle");
exports.TwoAssetRatioOracle = TwoAssetRatioOracle_1.TwoAssetRatioOracle;
var LegacyMakerOracleAdapter_1 = require("./ts/LegacyMakerOracleAdapter");
exports.LegacyMakerOracleAdapter = LegacyMakerOracleAdapter_1.LegacyMakerOracleAdapter;
var LinearizedEMATimeSeriesFeed_1 = require("./ts/LinearizedEMATimeSeriesFeed");
exports.LinearizedEMATimeSeriesFeed = LinearizedEMATimeSeriesFeed_1.LinearizedEMATimeSeriesFeed;
var LinearizedPriceDataSource_1 = require("./ts/LinearizedPriceDataSource");
exports.LinearizedPriceDataSource = LinearizedPriceDataSource_1.LinearizedPriceDataSource;
var Median_1 = require("./ts/Median");
exports.Median = Median_1.Median;
var MovingAverageOracle_1 = require("./ts/MovingAverageOracle");
exports.MovingAverageOracle = MovingAverageOracle_1.MovingAverageOracle;
var MovingAverageOracleV1Proxy_1 = require("./ts/MovingAverageOracleV1Proxy");
exports.MovingAverageOracleV1Proxy = MovingAverageOracleV1Proxy_1.MovingAverageOracleV1Proxy;
var MovingAverageOracleV2_1 = require("./ts/MovingAverageOracleV2");
exports.MovingAverageOracleV2 = MovingAverageOracleV2_1.MovingAverageOracleV2;
var OracleProxy_1 = require("./ts/OracleProxy");
exports.OracleProxy = OracleProxy_1.OracleProxy;
var OracleProxyCaller_1 = require("./ts/OracleProxyCaller");
exports.OracleProxyCaller = OracleProxyCaller_1.OracleProxyCaller;
var PriceFeed_1 = require("./ts/PriceFeed");
exports.PriceFeed = PriceFeed_1.PriceFeed;
var RSIOracle_1 = require("./ts/RSIOracle");
exports.RSIOracle = RSIOracle_1.RSIOracle;
var TimeSeriesFeed_1 = require("./ts/TimeSeriesFeed");
exports.TimeSeriesFeed = TimeSeriesFeed_1.TimeSeriesFeed;
var TimeSeriesFeedV2_1 = require("./ts/TimeSeriesFeedV2");
exports.TimeSeriesFeedV2 = TimeSeriesFeedV2_1.TimeSeriesFeedV2;
var TimeSeriesFeedV2Mock_1 = require("./ts/TimeSeriesFeedV2Mock");
exports.TimeSeriesFeedV2Mock = TimeSeriesFeedV2Mock_1.TimeSeriesFeedV2Mock;
var TwoAssetLinearizedTimeSeriesFeed_1 = require("./ts/TwoAssetLinearizedTimeSeriesFeed");
exports.TwoAssetLinearizedTimeSeriesFeed = TwoAssetLinearizedTimeSeriesFeed_1.TwoAssetLinearizedTimeSeriesFeed;
var UpdatableOracleMock_1 = require("./ts/UpdatableOracleMock");
exports.UpdatableOracleMock = UpdatableOracleMock_1.UpdatableOracleMock;
// Export abi-gen contract wrappers
var contracts_1 = require("../utils/contracts");
exports.BaseContract = contracts_1.BaseContract;
exports.ChainlinkAggregatorMockContract = contracts_1.ChainlinkAggregatorMockContract;
exports.ChainlinkOracleAdapterContract = contracts_1.ChainlinkOracleAdapterContract;
exports.ConstantPriceOracleContract = contracts_1.ConstantPriceOracleContract;
exports.CTokenOracleContract = contracts_1.CTokenOracleContract;
exports.DydxOracleAdapterContract = contracts_1.DydxOracleAdapterContract;
exports.DydxConstantPriceOracleMockContract = contracts_1.DydxConstantPriceOracleMockContract;
exports.EMAOracleContract = contracts_1.EMAOracleContract;
exports.FeedFactoryContract = contracts_1.FeedFactoryContract;
exports.HistoricalPriceFeedContract = contracts_1.HistoricalPriceFeedContract;
exports.TwoAssetRatioOracleContract = contracts_1.TwoAssetRatioOracleContract;
exports.LegacyMakerOracleAdapterContract = contracts_1.LegacyMakerOracleAdapterContract;
exports.LinkedListLibraryMockContract = contracts_1.LinkedListLibraryMockContract;
exports.LinearizedEMATimeSeriesFeedContract = contracts_1.LinearizedEMATimeSeriesFeedContract;
exports.LinearizedPriceDataSourceContract = contracts_1.LinearizedPriceDataSourceContract;
exports.MedianContract = contracts_1.MedianContract;
exports.MovingAverageOracleContract = contracts_1.MovingAverageOracleContract;
exports.MovingAverageOracleV1ProxyContract = contracts_1.MovingAverageOracleV1ProxyContract;
exports.MovingAverageOracleV2Contract = contracts_1.MovingAverageOracleV2Contract;
exports.OracleProxyContract = contracts_1.OracleProxyContract;
exports.OracleProxyCallerContract = contracts_1.OracleProxyCallerContract;
exports.PriceFeedContract = contracts_1.PriceFeedContract;
exports.RSIOracleContract = contracts_1.RSIOracleContract;
exports.TimeSeriesFeedContract = contracts_1.TimeSeriesFeedContract;
exports.TimeSeriesFeedV2Contract = contracts_1.TimeSeriesFeedV2Contract;
exports.TimeSeriesFeedV2MockContract = contracts_1.TimeSeriesFeedV2MockContract;
exports.TwoAssetLinearizedTimeSeriesFeedContract = contracts_1.TwoAssetLinearizedTimeSeriesFeedContract;
exports.UpdatableOracleMockContract = contracts_1.UpdatableOracleMockContract;
var oracleHelper_1 = require("../utils/helpers/oracleHelper");
exports.OracleHelper = oracleHelper_1.OracleHelper;
