// @0xproject/json-schemas/lib/src/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var basic_type_schemas_1 = require("../schemas/basic_type_schemas");
var block_range_schema_1 = require("../schemas/block_range_schema");
var call_data_schema_1 = require("../schemas/call_data_schema");
var ec_signature_schema_1 = require("../schemas/ec_signature_schema");
var index_filter_values_schema_1 = require("../schemas/index_filter_values_schema");
var order_cancel_schema_1 = require("../schemas/order_cancel_schema");
var order_fill_or_kill_requests_schema_1 = require("../schemas/order_fill_or_kill_requests_schema");
var order_fill_requests_schema_1 = require("../schemas/order_fill_requests_schema");
var order_hash_schema_1 = require("../schemas/order_hash_schema");
var order_schemas_1 = require("../schemas/order_schemas");
var orders_schema_1 = require("../schemas/orders_schema");
var paginated_collection_schema_1 = require("../schemas/paginated_collection_schema");
var relayer_api_asset_pairs_response_schema_1 = require("../schemas/relayer_api_asset_pairs_response_schema");
var relayer_api_error_response_schema_1 = require("../schemas/relayer_api_error_response_schema");
var relayer_api_fee_recipients_response_schema_1 = require("../schemas/relayer_api_fee_recipients_response_schema");
var relayer_api_order_config_payload_schema_1 = require("../schemas/relayer_api_order_config_payload_schema");
var relayer_api_order_config_response_schema_1 = require("../schemas/relayer_api_order_config_response_schema");
var relayer_api_order_schema_1 = require("../schemas/relayer_api_order_schema");
var relayer_api_orderbook_response_schema_1 = require("../schemas/relayer_api_orderbook_response_schema");
var relayer_api_orders_channel_subscribe_schema_1 = require("../schemas/relayer_api_orders_channel_subscribe_schema");
var relayer_api_orders_channel_update_response_schema_1 = require("../schemas/relayer_api_orders_channel_update_response_schema");
var relayer_api_orders_response_schema_1 = require("../schemas/relayer_api_orders_response_schema");
var relayer_api_orders_schema_1 = require("../schemas/relayer_api_orders_schema");
var signed_orders_schema_1 = require("../schemas/signed_orders_schema");
var token_schema_1 = require("../schemas/token_schema");
var tx_data_schema_1 = require("../schemas/tx_data_schema");
exports.schemas = {
    numberSchema: basic_type_schemas_1.numberSchema,
    addressSchema: basic_type_schemas_1.addressSchema,
    callDataSchema: call_data_schema_1.callDataSchema,
    hexSchema: basic_type_schemas_1.hexSchema,
    ecSignatureParameterSchema: ec_signature_schema_1.ecSignatureParameterSchema,
    ecSignatureSchema: ec_signature_schema_1.ecSignatureSchema,
    indexFilterValuesSchema: index_filter_values_schema_1.indexFilterValuesSchema,
    orderCancellationRequestsSchema: order_cancel_schema_1.orderCancellationRequestsSchema,
    orderFillOrKillRequestsSchema: order_fill_or_kill_requests_schema_1.orderFillOrKillRequestsSchema,
    orderFillRequestsSchema: order_fill_requests_schema_1.orderFillRequestsSchema,
    orderHashSchema: order_hash_schema_1.orderHashSchema,
    orderSchema: order_schemas_1.orderSchema,
    signedOrderSchema: order_schemas_1.signedOrderSchema,
    signedOrdersSchema: signed_orders_schema_1.signedOrdersSchema,
    ordersSchema: orders_schema_1.ordersSchema,
    blockParamSchema: block_range_schema_1.blockParamSchema,
    blockRangeSchema: block_range_schema_1.blockRangeSchema,
    tokenSchema: token_schema_1.tokenSchema,
    jsNumber: tx_data_schema_1.jsNumber,
    txDataSchema: tx_data_schema_1.txDataSchema,
    paginatedCollectionSchema: paginated_collection_schema_1.paginatedCollectionSchema,
    relayerApiErrorResponseSchema: relayer_api_error_response_schema_1.relayerApiErrorResponseSchema,
    relayerApiFeeRecipientsResponseSchema: relayer_api_fee_recipients_response_schema_1.relayerApiFeeRecipientsResponseSchema,
    relayerApiOrderSchema: relayer_api_order_schema_1.relayerApiOrderSchema,
    relayerApiOrdersSchema: relayer_api_orders_schema_1.relayerApiOrdersSchema,
    relayerApiOrderConfigPayloadSchema: relayer_api_order_config_payload_schema_1.relayerApiOrderConfigPayloadSchema,
    relayerApiOrderConfigResponseSchema: relayer_api_order_config_response_schema_1.relayerApiOrderConfigResponseSchema,
    relayerApiOrderbookResponseSchema: relayer_api_orderbook_response_schema_1.relayerApiOrderbookResponseSchema,
    relayerApiAssetDataPairsResponseSchema: relayer_api_asset_pairs_response_schema_1.relayerApiAssetDataPairsResponseSchema,
    relayerApiAssetDataTradeInfoSchema: relayer_api_asset_pairs_response_schema_1.relayerApiAssetDataTradeInfoSchema,
    relayerApiOrdersChannelSubscribeSchema: relayer_api_orders_channel_subscribe_schema_1.relayerApiOrdersChannelSubscribeSchema,
    relayerApiOrdersChannelSubscribePayload: relayer_api_orders_channel_subscribe_schema_1.relayerApiOrdersChannelSubscribePayload,
    relayerApiOrdersChannelUpdateSchema: relayer_api_orders_channel_update_response_schema_1.relayerApiOrdersChannelUpdateSchema,
    relayerApiOrdersResponseSchema: relayer_api_orders_response_schema_1.relayerApiOrdersResponseSchema,
    relayerApiAssetDataPairsSchema: relayer_api_asset_pairs_response_schema_1.relayerApiAssetDataPairsSchema,
};
