// set-protocol-contracts/dist/types/generated/core_module_interaction.js
/*
 Copyright 2018 Set Labs Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
* This file is auto-generated using abi-gen. Don't edit directly.
* Templates can be found at https://github.com/0xProject/0x.js/tree/development/packages/abi-gen-templates.
*/
var utils_1 = require("@0xproject/utils");
var set_protocol_utils_1 = require("set-protocol-utils");
var base_contract_1 = require("../base_contract");
var common_1 = require("../common");
var CoreModuleInteractionContract = /** @class */ (function (_super) {
    __extends(CoreModuleInteractionContract, _super);
    function CoreModuleInteractionContract(web3ContractInstance, defaults) {
        var _this = _super.call(this, web3ContractInstance, defaults) || this;
        _this.validFactories = {
            callAsync: function (_factory, defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.validFactories(this.formatABIDataItem('address', [], _factory)).call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.exchanges = {
            callAsync: function (defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.exchanges().call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.setTokens = {
            callAsync: function (defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.setTokens().call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.exchangeIds = {
            callAsync: function (_exchangeId, defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.exchangeIds(this.formatABIDataItem('uint8', [], _exchangeId)).call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.validModules = {
            callAsync: function (_module, defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.validModules(this.formatABIDataItem('address', [], _module)).call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.transferProxy = {
            callAsync: function (defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.transferProxy().call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.operationState = {
            callAsync: function (defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.operationState().call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, new set_protocol_utils_1.BigNumber(result)];
                        }
                    });
                });
            },
        };
        _this.validPriceLibraries = {
            callAsync: function (_priceLibrary, defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.validPriceLibraries(this.formatABIDataItem('address', [], _priceLibrary)).call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.state = {
            callAsync: function (defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.state().call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.priceLibraries = {
            callAsync: function (defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.priceLibraries().call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.disabledSets = {
            callAsync: function (_set, defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.disabledSets(this.formatABIDataItem('address', [], _set)).call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.modules = {
            callAsync: function (defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.modules().call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.vault = {
            callAsync: function (defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.vault().call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.factories = {
            callAsync: function (defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.factories().call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.validSets = {
            callAsync: function (_set, defaultBlock) {
                return __awaiter(this, void 0, void 0, function () {
                    var self, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.validSets(this.formatABIDataItem('address', [], _set)).call, self.web3ContractInstance)()];
                            case 1:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.depositModule = {
            sendTransactionAsync: function (_from, _to, _token, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.depositModule.estimateGasAsync.bind(self, this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.depositModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_from, _to, _token, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.depositModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_from, _to, _token, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.depositModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_from, _to, _token, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.depositModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.batchDepositModule = {
            sendTransactionAsync: function (_from, _to, _tokens, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.batchDepositModule.estimateGasAsync.bind(self, this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchDepositModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_from, _to, _tokens, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchDepositModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_from, _to, _tokens, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.batchDepositModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_from, _to, _tokens, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchDepositModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.withdrawModule = {
            sendTransactionAsync: function (_from, _to, _token, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.withdrawModule.estimateGasAsync.bind(self, this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.withdrawModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_from, _to, _token, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.withdrawModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_from, _to, _token, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.withdrawModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_from, _to, _token, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.withdrawModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.batchWithdrawModule = {
            sendTransactionAsync: function (_from, _to, _tokens, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.batchWithdrawModule.estimateGasAsync.bind(self, this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchWithdrawModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_from, _to, _tokens, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchWithdrawModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_from, _to, _tokens, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.batchWithdrawModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_from, _to, _tokens, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchWithdrawModule(this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.issueModule = {
            sendTransactionAsync: function (_componentOwner, _setRecipient, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.issueModule.estimateGasAsync.bind(self, this.formatABIDataItem('address', [], _componentOwner), this.formatABIDataItem('address', [], _setRecipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.issueModule(this.formatABIDataItem('address', [], _componentOwner), this.formatABIDataItem('address', [], _setRecipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_componentOwner, _setRecipient, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.issueModule(this.formatABIDataItem('address', [], _componentOwner), this.formatABIDataItem('address', [], _setRecipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_componentOwner, _setRecipient, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.issueModule(this.formatABIDataItem('address', [], _componentOwner), this.formatABIDataItem('address', [], _setRecipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_componentOwner, _setRecipient, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.issueModule(this.formatABIDataItem('address', [], _componentOwner), this.formatABIDataItem('address', [], _setRecipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.issueInVaultModule = {
            sendTransactionAsync: function (_recipient, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.issueInVaultModule.estimateGasAsync.bind(self, this.formatABIDataItem('address', [], _recipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.issueInVaultModule(this.formatABIDataItem('address', [], _recipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_recipient, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.issueInVaultModule(this.formatABIDataItem('address', [], _recipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_recipient, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.issueInVaultModule(this.formatABIDataItem('address', [], _recipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_recipient, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.issueInVaultModule(this.formatABIDataItem('address', [], _recipient), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.redeemModule = {
            sendTransactionAsync: function (_burnAddress, _incrementAddress, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.redeemModule.estimateGasAsync.bind(self, this.formatABIDataItem('address', [], _burnAddress), this.formatABIDataItem('address', [], _incrementAddress), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.redeemModule(this.formatABIDataItem('address', [], _burnAddress), this.formatABIDataItem('address', [], _incrementAddress), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_burnAddress, _incrementAddress, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.redeemModule(this.formatABIDataItem('address', [], _burnAddress), this.formatABIDataItem('address', [], _incrementAddress), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_burnAddress, _incrementAddress, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.redeemModule(this.formatABIDataItem('address', [], _burnAddress), this.formatABIDataItem('address', [], _incrementAddress), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_burnAddress, _incrementAddress, _set, _quantity, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.redeemModule(this.formatABIDataItem('address', [], _burnAddress), this.formatABIDataItem('address', [], _incrementAddress), this.formatABIDataItem('address', [], _set), this.formatABIDataItem('uint256', [], _quantity)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.batchIncrementTokenOwnerModule = {
            sendTransactionAsync: function (_tokens, _owner, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.batchIncrementTokenOwnerModule.estimateGasAsync.bind(self, this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchIncrementTokenOwnerModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_tokens, _owner, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchIncrementTokenOwnerModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_tokens, _owner, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.batchIncrementTokenOwnerModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_tokens, _owner, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchIncrementTokenOwnerModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.batchDecrementTokenOwnerModule = {
            sendTransactionAsync: function (_tokens, _owner, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.batchDecrementTokenOwnerModule.estimateGasAsync.bind(self, this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchDecrementTokenOwnerModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_tokens, _owner, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchDecrementTokenOwnerModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_tokens, _owner, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.batchDecrementTokenOwnerModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_tokens, _owner, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchDecrementTokenOwnerModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _owner), this.formatABIDataItem('uint256[]', [], _quantities)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.batchTransferBalanceModule = {
            sendTransactionAsync: function (_tokens, _from, _to, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.batchTransferBalanceModule.estimateGasAsync.bind(self, this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('uint256[]', [], _quantities)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchTransferBalanceModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('uint256[]', [], _quantities)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_tokens, _from, _to, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchTransferBalanceModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('uint256[]', [], _quantities)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_tokens, _from, _to, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.batchTransferBalanceModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('uint256[]', [], _quantities)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_tokens, _from, _to, _quantities, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchTransferBalanceModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to), this.formatABIDataItem('uint256[]', [], _quantities)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.transferModule = {
            sendTransactionAsync: function (_token, _quantity, _from, _to, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.transferModule.estimateGasAsync.bind(self, this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.transferModule(this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_token, _quantity, _from, _to, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.transferModule(this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_token, _quantity, _from, _to, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.transferModule(this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_token, _quantity, _from, _to, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.transferModule(this.formatABIDataItem('address', [], _token), this.formatABIDataItem('uint256', [], _quantity), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        _this.batchTransferModule = {
            sendTransactionAsync: function (_tokens, _quantities, _from, _to, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, txHash;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData, self.batchTransferModule.estimateGasAsync.bind(self, this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)))];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchTransferModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)).send, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                txHash = _a.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_tokens, _quantities, _from, _to, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, gas;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchTransferModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)).estimateGas, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                gas = _a.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_tokens, _quantities, _from, _to, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var abiEncodedTransactionData = self.web3ContractInstance.methods.batchTransferModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)).encodeABI();
                return abiEncodedTransactionData;
            },
            callAsync: function (_tokens, _quantities, _from, _to, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, txDataWithDefaults, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                return [4 /*yield*/, self.applyDefaultsToTxDataAsync(txData)];
                            case 1:
                                txDataWithDefaults = _a.sent();
                                return [4 /*yield*/, utils_1.promisify(self.web3ContractInstance.methods.batchTransferModule(this.formatABIDataItem('address[]', [], _tokens), this.formatABIDataItem('uint256[]', [], _quantities), this.formatABIDataItem('address', [], _from), this.formatABIDataItem('address', [], _to)).call, self.web3ContractInstance)(txDataWithDefaults)];
                            case 2:
                                result = _a.sent();
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
        };
        common_1.classUtils.bindAll(_this, ['web3ContractInstance', 'defaults']);
        return _this;
    }
    CoreModuleInteractionContract.prototype.deploy = function (data, args) {
        return __awaiter(this, void 0, void 0, function () {
            var wrapper;
            return __generator(this, function (_a) {
                wrapper = this;
                wrapper.web3ContractInstance.deploy({
                    data: data,
                    arguments: args,
                })
                    .send(wrapper.defaults)
                    .then(function (newContractInstance) {
                    wrapper.web3ContractInstance = newContractInstance;
                    wrapper.address = newContractInstance.options.address;
                });
                return [2 /*return*/];
            });
        });
    };
    CoreModuleInteractionContract.deployed = function (web3, defaults) {
        return __awaiter(this, void 0, void 0, function () {
            var web3Utils, currentNetwork, _a, abi, networks, web3ContractInstance;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        web3Utils = new set_protocol_utils_1.Web3Utils(web3);
                        return [4 /*yield*/, web3Utils.getNetworkIdAsync()];
                    case 1:
                        currentNetwork = _b.sent();
                        _a = require('../../artifacts/ts/CoreModuleInteraction').CoreModuleInteraction, abi = _a.abi, networks = _a.networks;
                        web3ContractInstance = new web3.eth.Contract(abi, networks[currentNetwork].address);
                        return [2 /*return*/, new CoreModuleInteractionContract(web3ContractInstance, defaults)];
                }
            });
        });
    };
    CoreModuleInteractionContract.at = function (address, web3, defaults) {
        return __awaiter(this, void 0, void 0, function () {
            var abi, web3Utils, contractExists, currentNetwork, web3ContractInstance;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        abi = require('../../artifacts/ts/CoreModuleInteraction').CoreModuleInteraction.abi;
                        web3Utils = new set_protocol_utils_1.Web3Utils(web3);
                        return [4 /*yield*/, web3Utils.doesContractExistAtAddressAsync(address)];
                    case 1:
                        contractExists = _a.sent();
                        return [4 /*yield*/, web3Utils.getNetworkIdAsync()];
                    case 2:
                        currentNetwork = _a.sent();
                        if (contractExists) {
                            web3ContractInstance = new web3.eth.Contract(abi, address);
                            return [2 /*return*/, new CoreModuleInteractionContract(web3ContractInstance, defaults)];
                        }
                        else {
                            throw new Error(base_contract_1.CONTRACT_WRAPPER_ERRORS.CONTRACT_NOT_FOUND_ON_NETWORK('CoreModuleInteraction', currentNetwork));
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return CoreModuleInteractionContract;
}(base_contract_1.BaseContract)); // tslint:disable:max-file-line-count
exports.CoreModuleInteractionContract = CoreModuleInteractionContract;
