// set-protocol-oracles/dist/utils/contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var base_contract_1 = require("../types/base_contract");
exports.BaseContract = base_contract_1.BaseContract;
var chainlink_aggregator_mock_1 = require("../types/generated/chainlink_aggregator_mock");
exports.ChainlinkAggregatorMockContract = chainlink_aggregator_mock_1.ChainlinkAggregatorMockContract;
var chainlink_oracle_adapter_1 = require("../types/generated/chainlink_oracle_adapter");
exports.ChainlinkOracleAdapterContract = chainlink_oracle_adapter_1.ChainlinkOracleAdapterContract;
var constant_price_oracle_1 = require("../types/generated/constant_price_oracle");
exports.ConstantPriceOracleContract = constant_price_oracle_1.ConstantPriceOracleContract;
var c_token_oracle_1 = require("../types/generated/c_token_oracle");
exports.CTokenOracleContract = c_token_oracle_1.CTokenOracleContract;
var data_source_linear_interpolation_library_mock_1 = require("../types/generated/data_source_linear_interpolation_library_mock");
exports.DataSourceLinearInterpolationLibraryMockContract = data_source_linear_interpolation_library_mock_1.DataSourceLinearInterpolationLibraryMockContract;
var dydx_constant_price_oracle_mock_1 = require("../types/generated/dydx_constant_price_oracle_mock");
exports.DydxConstantPriceOracleMockContract = dydx_constant_price_oracle_mock_1.DydxConstantPriceOracleMockContract;
var dydx_oracle_adapter_1 = require("../types/generated/dydx_oracle_adapter");
exports.DydxOracleAdapterContract = dydx_oracle_adapter_1.DydxOracleAdapterContract;
var e_m_a_library_mock_1 = require("../types/generated/e_m_a_library_mock");
exports.EMALibraryMockContract = e_m_a_library_mock_1.EMALibraryMockContract;
var e_m_a_oracle_1 = require("../types/generated/e_m_a_oracle");
exports.EMAOracleContract = e_m_a_oracle_1.EMAOracleContract;
var feed_factory_1 = require("../types/generated/feed_factory");
exports.FeedFactoryContract = feed_factory_1.FeedFactoryContract;
var historical_price_feed_1 = require("../types/generated/historical_price_feed");
exports.HistoricalPriceFeedContract = historical_price_feed_1.HistoricalPriceFeedContract;
var two_asset_ratio_oracle_1 = require("../types/generated/two_asset_ratio_oracle");
exports.TwoAssetRatioOracleContract = two_asset_ratio_oracle_1.TwoAssetRatioOracleContract;
var legacy_maker_oracle_adapter_1 = require("../types/generated/legacy_maker_oracle_adapter");
exports.LegacyMakerOracleAdapterContract = legacy_maker_oracle_adapter_1.LegacyMakerOracleAdapterContract;
var linked_list_helper_mock_1 = require("../types/generated/linked_list_helper_mock");
exports.LinkedListHelperMockContract = linked_list_helper_mock_1.LinkedListHelperMockContract;
var linked_list_library_mock_1 = require("../types/generated/linked_list_library_mock");
exports.LinkedListLibraryMockContract = linked_list_library_mock_1.LinkedListLibraryMockContract;
var linked_list_library_mock_v2_1 = require("../types/generated/linked_list_library_mock_v2");
exports.LinkedListLibraryMockV2Contract = linked_list_library_mock_v2_1.LinkedListLibraryMockV2Contract;
var linked_list_library_mock_v3_1 = require("../types/generated/linked_list_library_mock_v3");
exports.LinkedListLibraryMockV3Contract = linked_list_library_mock_v3_1.LinkedListLibraryMockV3Contract;
var linearized_e_m_a_time_series_feed_1 = require("../types/generated/linearized_e_m_a_time_series_feed");
exports.LinearizedEMATimeSeriesFeedContract = linearized_e_m_a_time_series_feed_1.LinearizedEMATimeSeriesFeedContract;
var linearized_price_data_source_1 = require("../types/generated/linearized_price_data_source");
exports.LinearizedPriceDataSourceContract = linearized_price_data_source_1.LinearizedPriceDataSourceContract;
var median_1 = require("../types/generated/median");
exports.MedianContract = median_1.MedianContract;
var moving_average_oracle_1 = require("../types/generated/moving_average_oracle");
exports.MovingAverageOracleContract = moving_average_oracle_1.MovingAverageOracleContract;
var moving_average_oracle_v2_1 = require("../types/generated/moving_average_oracle_v2");
exports.MovingAverageOracleV2Contract = moving_average_oracle_v2_1.MovingAverageOracleV2Contract;
var moving_average_oracle_v1_proxy_1 = require("../types/generated/moving_average_oracle_v1_proxy");
exports.MovingAverageOracleV1ProxyContract = moving_average_oracle_v1_proxy_1.MovingAverageOracleV1ProxyContract;
var oracle_proxy_caller_1 = require("../types/generated/oracle_proxy_caller");
exports.OracleProxyCallerContract = oracle_proxy_caller_1.OracleProxyCallerContract;
var oracle_proxy_1 = require("../types/generated/oracle_proxy");
exports.OracleProxyContract = oracle_proxy_1.OracleProxyContract;
var price_feed_1 = require("../types/generated/price_feed");
exports.PriceFeedContract = price_feed_1.PriceFeedContract;
var price_feed_mock_1 = require("../types/generated/price_feed_mock");
exports.PriceFeedMockContract = price_feed_mock_1.PriceFeedMockContract;
var r_s_i_library_mock_1 = require("../types/generated/r_s_i_library_mock");
exports.RSILibraryMockContract = r_s_i_library_mock_1.RSILibraryMockContract;
var r_s_i_oracle_1 = require("../types/generated/r_s_i_oracle");
exports.RSIOracleContract = r_s_i_oracle_1.RSIOracleContract;
var standard_token_mock_1 = require("../types/generated/standard_token_mock");
exports.StandardTokenMockContract = standard_token_mock_1.StandardTokenMockContract;
var time_series_feed_1 = require("../types/generated/time_series_feed");
exports.TimeSeriesFeedContract = time_series_feed_1.TimeSeriesFeedContract;
var time_series_feed_v2_1 = require("../types/generated/time_series_feed_v2");
exports.TimeSeriesFeedV2Contract = time_series_feed_v2_1.TimeSeriesFeedV2Contract;
var time_series_feed_v2_mock_1 = require("../types/generated/time_series_feed_v2_mock");
exports.TimeSeriesFeedV2MockContract = time_series_feed_v2_mock_1.TimeSeriesFeedV2MockContract;
var two_asset_linearized_time_series_feed_1 = require("../types/generated/two_asset_linearized_time_series_feed");
exports.TwoAssetLinearizedTimeSeriesFeedContract = two_asset_linearized_time_series_feed_1.TwoAssetLinearizedTimeSeriesFeedContract;
var updatable_oracle_mock_1 = require("../types/generated/updatable_oracle_mock");
exports.UpdatableOracleMockContract = updatable_oracle_mock_1.UpdatableOracleMockContract;
var u_s_d_c_mock_1 = require("../types/generated/u_s_d_c_mock");
exports.USDCMockContract = u_s_d_c_mock_1.USDCMockContract;
var weth_mock_1 = require("../types/generated/weth_mock");
exports.WethMockContract = weth_mock_1.WethMockContract;
